src/aliceVision/depthMap/cuda/commonStructures.hpp

```diff
@@ -131,441 +155,748 @@ CudaSize<Dim> operator-(const CudaSize<Dim> &lhs, const CudaSize<Dim> &rhs) {
 {
   cudaMemcpyKind kind = cudaMemcpyDeviceToHost;
-  cudaError_t err = cudaMemcpy(_dst.getBuffer(), _src.getBuffer(), _src.getBytes(), kind);
+  cudaError_t err = cudaMemcpy(_dst.getBytePtr(),
+                               _src.getBytePtr(),
+                               _dst.getUnpaddedBytesInDim(0),
+                               kind);
   THROW_ON_CUDA_ERROR(err, "Failed to copy from CudaHostMemoryHeap to CudaDeviceMemory");
 }
 
 template<class Type, unsigned Dim> void copy(CudaHostMemoryHeap<Type, Dim>& _dst, const CudaArray<Type, Dim>& _src)
 {
   cudaMemcpyKind kind = cudaMemcpyDeviceToHost;
   if(Dim == 1) {
-    cudaError_t err = cudaMemcpyFromArray(_dst.getBuffer(), _src.getArray(), 0, 0, _dst.getSize()[0] * sizeof (Type), kind);
+    cudaError_t err = cudaMemcpyFromArray(_dst.getBytePtr(),
+                                          _src.getArray(),
+                                          0, 0,
+                                          _dst.getUnpaddedBytesInDim(0),
+                                          kind);
     THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
-  else if(Dim == 2) {
-    cudaError_t err = cudaMemcpy2DFromArray(_dst.getBuffer(),
-                                            _dst.getSize()[0] * sizeof (Type),
+  else if(Dim == 2)
+  {
+    cudaError_t err = cudaMemcpy2DFromArray(_dst.getBytePtr(),
+                                            _dst.getPitch(),
                                             _src.getArray(),
                                             0,
                                             0,
-                                            _dst.getSize()[0] * sizeof (Type),
-                                            _dst.getSize()[1],
+                                            _dst.getUnpaddedBytesInDim(0),
+                                            _dst.getUnpaddedBytesInDim(1),
                                             kind);
     THROW_ON_CUDA_ERROR(err, "Failed to copy (" << __FILE__ << " " << __LINE__ << ")");
   }
-  else if(Dim == 3) {
+  else if(Dim >= 3)
+  {
+    size_t number_of_rows = 1;
+    for( int i=1; i<Dim; i++ ) number_of_rows *= _dst.getUnitsInDim(i);
+
     cudaMemcpy3DParms p = { 0 };
     p.srcArray = const_cast<cudaArray *>(_src.getArray());
-    p.srcPtr.pitch = _src.getPitch();
-    p.srcPtr.xsize = _src.getSize()[0];
-    p.srcPtr.ysize = _src.getSize()[1];
-    p.dstPtr.ptr = (void *)_dst.getBuffer();
-    p.dstPtr.pitch = _dst.getSize()[0] * sizeof (Type);
-    p.dstPtr.xsize = _dst.getSize()[0];
-    p.dstPtr.ysize = _dst.getSize()[1];
-    p.extent.width = _dst.getSize()[0];
-    p.extent.height = _dst.getSize()[1];
-    p.extent.depth = _dst.getSize()[2];
+
+    p.dstPtr.ptr = (void *)_dst.getBytePtr();
+    p.dstPtr.pitch  = _dst.getPitch();
+    p.dstPtr.xsize  = _dst.getUnitsInDim(0);
+    p.dstPtr.ysize  = number_of_rows;
```

Comment:
When Dim is 3:
`p.dstPtr.ysize` is now `_dst.getUnitsInDim(1) * _dst.getUnitsInDim(2)` while it was: `_dst.getUnitsInDim(1)` before.
Is that right?